### ISSUE TITLE
Some topics are not consumed anymore

### ISSUE THREAD

**Issue by patrikni** [confluentinc/kafka-connect-hdfs#228]
Hi, 
I have a distributed deployment of Kafka Connect 3.2.1. There are about 80 topics with 10 partitions each. 
All topics are stored as HDFS files with HIVE integration turned on.  Topics are consumed in 10 single thread connectors. 
The strange is that some topics are not consumed at all. When HDFS directory is empty it consume one partition with 10 avro files. Nevertheless the next partitions are not created. If I recreate the connectors configuration using REST API. It start without problem but it does not consume. If I check last kafka offsets it is much bigger than last offset stored in avro files.
Other topics are consumed properly and the offset difference is very small. 

The connector still logs (the connector was running more than 30 minutes):
2017-09-19 17:25:11,687 INFO  [pool-1-thread-19] Starting commit and rotation for topic partition ACCOUNT_CORE-0 with start offsets {} and end offsets {} (io.confluent.connect.hdfs.TopicPartitionWriter)
2017-09-19 17:25:13,037 INFO  [pool-1-thread-19] WorkerSinkTask{id=sbd-sink} Committing offsets (org.apache.kafka.connect.runtime.WorkerSinkTask)

The connector configuration.
{
    "config": {
        "connect.hdfs.keytab": "/home/usr_kfk/usr_kfk.keytab",
        "connect.hdfs.principal": "usr_kfk@COMPANY.COM",
        "connector.class": "io.confluent.connect.hdfs.HdfsSinkConnector",
        "flush.size": "10000000",
        "format.class": "io.confluent.connect.hdfs.avro.AvroFormat",
        "hadoop.conf.dir": "/etc/hadoop/conf",
        "hdfs.authentication.kerberos": "true",
        "hdfs.namenode.principal": "hdfs/_HOST@COMPANY.COM",
        "hdfs.url": "hdfs://HOST-ns",
        "hive.conf.dir": "/etc/hive/conf",
        "hive.database": "kafkadb",
        "hive.integration": "false",
        "hive.metastore.uris": "thrift://srvhdp04:9083",
        "kerberos.ticket.renew.period.ms": "3600000",        
        "logs.dir": "/data/speed/logs",
        "name": "sbd-sink",
        "partition.duration.ms": "300000",
        "partitioner.class": "io.confluent.connect.hdfs.partitioner.TimeBasedPartitioner",
        "path.format": "'year'=YYYY/'month'=MM/'day'=dd/'hour'=HH/'minute'=mm/",
        "rotate.schedule.interval.ms": "300000",
        "schema.compatibility": "BACKWARD",
        "tasks.max": "10",
        "topics": "ACCOUNT_CORE",
        "topics.dir": "/data/speed/topics"
    },

**Comment by patrikni** [comment]
It seems that it is caused by missing record offsets. The TopicPartitionWriter gets recors by the  "public void buffer(SinkRecord sinkRecord)"  method. There are missing offsets in records. For example I have got offsets for partition 7 of my topic as follows  2521232, **2521234**, 2521235, 2521236, ... There is offset  2521234 before the offset 2521232! Message with the offset 2521233 is missing. 

The HDFS connector requres the offset order without gaps see TopicPartitionWriter.writeRecord(SinkRecord record). The methods ignores rest of messages after the offset gap.  
What is wrong? 
  1) Presumption that offset is without gaps ?
  2) HDFS connector implementation ?
  3) Kafka server - it has to deliver messages without gaps in topic partition ?

Thank for comments.

**Comment by patrikni** [comment]
Based on Internet search, missing offsets in a topic are possible for example in compacled topic. (https://stackoverflow.com/questions/43405503/repeatedly-produced-to-apache-kafka-different-offsets-exactly-once-semantics).  It is a BUG  of the HDFS connector.

**Comment by ewencp** [comment]
@patrikni This was fixed in https://github.com/confluentinc/kafka-connect-hdfs/pull/212 The check was originally added because of a bug in the Kafka Connect framework that could allow for incorrect records to be delivered to the connector. As that underlying bug had been fixed 2 major versions ago, we removed this check and the connector should now work fine for compacted topics. The check was removed in master, so the fix will be included with the next major release of the platform.
